robot_3d/src/joint_states_subscriber_3d.cpp

```diff
@@ -0,0 +1,233 @@
+  position[2] = msg->position[2];
+  position[3] = msg->position[3];
+  position[4] = msg->position[4];
+  position[5] = msg->position[5];
+
+  // Passing joint positons to method to calculate end effector position.
+  kdlCalc(position);
+
+  // Terminates node if the maximum number of data points is achieved.
+  if (data_point_count_ == num_data_point_)
+  {
+    ros::shutdown();
+  }
+}
+
+void JointMeasurementData::kdlCalc(double joint_positions[])
+{
+  // Constructing kdl tree from robot urdf.
+  KDL::Tree fanuc_tree;
+  std::string fanuc_robot;
+
+  ros::NodeHandle n;
+  n.param("/robot_description", fanuc_robot, std::string());
+  if (!kdl_parser::treeFromString(fanuc_robot, fanuc_tree))
+  {
+    ROS_ERROR("Failed to construct kdl tree");
+  }
+
+  // Creating kdl chain from base link to end effector.
+  KDL::Chain chain;
+  fanuc_tree.getChain("base_link", "tool0", chain);
+
+  KDL::ChainFkSolverPos_recursive fksolver = KDL::ChainFkSolverPos_recursive(chain);
+
+  // Creating a kdl array to store joint positions.
+  unsigned int nj = chain.getNrOfJoints();
+  KDL::JntArray jointpositions = KDL::JntArray(nj);
+  ROS_INFO("NrOfJoints =%d", nj);
+
+  // assign joint positions...
+  for (unsigned int i = 0; i < nj; i++)
+  {
+    jointpositions(i) = joint_positions[i];
+  }
+
+  // Create the frame that will contain the end effector pose.
+  KDL::Frame cartpos;
+
+  // Calculate forward kinematics for end effector.
+  bool kinematics_status;
+  kinematics_status = fksolver.JntToCart(jointpositions, cartpos);
+  if (kinematics_status >= 0)
+  {
+    std::cout << cartpos << std::endl;
+    ROS_INFO("%s \n", "Success, thanks KDL!");
+  }
+  else
+  {
+    ROS_INFO("%s \n", "Error: could not calculate forward kinematics :(");
+  }
+
+  // Creating and broadcasting tf transform for end effector.
+  tf::Transform transform;
+  tf::transformKDLToTF(cartpos, transform);
+
+  static tf::TransformBroadcaster br;
+  br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), "base_link", "test_frame"));
+
+  // Outputting data to yaml file.
+  outputYaml(transform, joint_positions);
+}
+
+std::vector<double> JointMeasurementData::getJointOffsets(double joint_positions[])
+{
+  ros::NodeHandle n;
+
+  // Vector to store the offset joint angles for robot.
+  std::vector<double> offset_joints;
+
+  n.getParam("offsets", offset_joints);
+
+  // Adding offsets to joint angles.
+  for (unsigned i = 0; i < offset_joints.size(); i++)
+  {
+    offset_joints[i] += joint_positions[i];
+  }
+
+  return offset_joints;
+}
+
+void JointMeasurementData::outputYaml(tf::Transform kdl_transform, double joint_positions[])
+{
+  // Storing the position and orientation of end effector.
+  tf::Vector3 translation = kdl_transform.getOrigin();
+  tf::Quaternion orientation = kdl_transform.getRotation();
+
+  // Offset joint angles.
+  std::vector<double> offset_joint_angles = getJointOffsets(joint_positions);
+
+  // Outputting yaml stuff
+  std::stringstream stream;
```

Comment:
I know I said to use a string steam if you were building the YAML output as a string. But since you are using the YAML emitter, is it possible to output directly to the file ofstream?

Reply:
If not then thats fine too, just LMK.